EventFilter/HGCalRawToDigi/plugins/HGCalRawToDigi.cc

```diff
@@ -0,0 +1,130 @@
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  const edm::EDGetTokenT<FEDRawDataCollection> fedRawToken_;
+  const edm::EDPutTokenT<HGCalDigiCollection> digisToken_;
+  const edm::EDPutTokenT<HGCalElecDigiCollection> elecDigisToken_;
+
+  const std::vector<unsigned int> fedIds_;
+  const unsigned int badECONDMax_;
+  const unsigned int numERxsInECOND_;
+  const std::unique_ptr<HGCalUnpacker<HGCalElectronicsId> > unpacker_;
+};
+
+HGCalRawToDigi::HGCalRawToDigi(const edm::ParameterSet& iConfig)
+    : fedRawToken_(consumes<FEDRawDataCollection>(iConfig.getParameter<edm::InputTag>("src"))),
+      digisToken_(produces<HGCalDigiCollection>()),
+      elecDigisToken_(produces<HGCalElecDigiCollection>()),
+      fedIds_(iConfig.getParameter<std::vector<unsigned int> >("fedIds")),
+      badECONDMax_(iConfig.getParameter<unsigned int>("badECONDMax")),
+      numERxsInECOND_(iConfig.getParameter<unsigned int>("numERxsInECOND")),
+      unpacker_(new HGCalUnpacker<HGCalElectronicsId>(
+          HGCalUnpackerConfig{.sLinkBOE = iConfig.getParameter<unsigned int>("slinkBOE"),
+                              .captureBlockReserved = iConfig.getParameter<unsigned int>("captureBlockReserved"),
+                              .econdHeaderMarker = iConfig.getParameter<unsigned int>("econdHeaderMarker"),
+                              .sLinkCaptureBlockMax = iConfig.getParameter<unsigned int>("maxCaptureBlock"),
+                              .captureBlockECONDMax = iConfig.getParameter<unsigned int>("captureBlockECONDMax"),
+                              .econdERXMax = iConfig.getParameter<unsigned int>("econdERXMax"),
+                              .erxChannelMax = iConfig.getParameter<unsigned int>("erxChannelMax"),
+                              .payloadLengthMax = iConfig.getParameter<unsigned int>("payloadLengthMax"),
+                              .channelMax = iConfig.getParameter<unsigned int>("channelMax"),
+                              .commonModeMax = iConfig.getParameter<unsigned int>("commonModeMax")})) {}
+
+void HGCalRawToDigi::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  // retrieve the FED raw data
+  const auto& raw_data = iEvent.get(fedRawToken_);
+  // prepare the output
+  HGCalDigiCollection digis;
+  HGCalElecDigiCollection elec_digis;
+  for (const auto& fed_id : fedIds_) {
+    const auto& fed_data = raw_data.FEDData(fed_id);
+    if (fed_data.size() == 0)
+      continue;
+
+    std::vector<uint32_t> data_32bit;
+    auto* ptr = fed_data.data();
+    for (size_t i = 0; i < fed_data.size(); i += 4)
+      data_32bit.emplace_back(((ptr + i) ? ((*(ptr + i) & 0xff) << 0) : 0) +
+                              ((ptr + i + 1) ? ((*(ptr + i + 1) & 0xff) << 8) : 0) +
+                              ((ptr + i + 2) ? ((*(ptr + i + 2) & 0xff) << 16) : 0) +
+                              ((ptr + i + 3) ? ((*(ptr + i + 3) & 0xff) << 24) : 0));
```

Comment:
@forthommel the el8_amd64_gcc12 builds are complaining in the IB about these lines, see e.g. https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_13_2_X_2023-05-27-1100/EventFilter/HGCalRawToDigi

Indeed: what are you checking here? The pointer always exists, it may point to some non existent field at most.

Could something as follows correspond to what you had in mind, maybe?
```C++
    size_t fed_data_size = fed_data.size();
    for (size_t i = 0; i < fed_data_size; i += 4)
      data_32bit.emplace_back(((*(ptr + i) & 0xff) << 0)  +
                              ((i + 1 < fed_data_size )? (*(ptr + i + 1) & 0xff) << 8) : 0) +
                              ((i + 2 < fed_data_size) ? ((*(ptr + i + 2) & 0xff) << 16) : 0) +
                              ((i + 3 < fed_data_size) ? ((*(ptr + i + 3) & 0xff) << 24) : 0));
```

Reply:
Hi @perrotta !
Thanks for the extra x-check. Sure, this would indeed be a better practice. I can prepare a PR candidate to follow this up.
As a matter of fact I was actually surprised such a simple words merging operation was not already part of the common `FWCore/Utilities` tools!